src/soter/soter_container.c

```diff
@@ -35,11 +34,31 @@ soter_status_t soter_verify_container_checksum(const soter_container_hdr_t* hdr)
 
     soter_crc32_update(&crc, hdr, sizeof(soter_container_hdr_t) - sizeof(uint32_t));
     soter_crc32_update(&crc, &dummy_crc, sizeof(uint32_t));
-    soter_crc32_update(&crc, hdr + 1, ntohl(hdr->size) - sizeof(soter_container_hdr_t));
+    soter_crc32_update(&crc, hdr + 1, be32toh(hdr->size) - sizeof(soter_container_hdr_t));
 
-    if (hdr->crc == htonl(soter_crc32_final(&crc))) {
+    if (hdr->crc == htobe32(soter_crc32_final(&crc))) {
         return SOTER_SUCCESS;
     }
 
     return SOTER_DATA_CORRUPT;
 }
+
+uint8_t* soter_container_data(soter_container_hdr_t* hdr)
+{
+    return (uint8_t*)(hdr + 1);
+}
+
+const uint8_t* soter_container_const_data(const soter_container_hdr_t* hdr)
+{
+    return (const uint8_t*)(hdr + 1);
+}
+
+size_t soter_container_data_size(const soter_container_hdr_t* hdr)
+{
+    return be32toh(hdr->size) - sizeof(soter_container_hdr_t);
```

Comment:
i feel insecure about these computations 🤔

Reply:
htr->size contains size of message including header itself. this computation return length of encrypted message without header. because size we pack as big-endian number, to we convert it to host endianness from big-endian;

Reply:
Well... These computations are copied verbatim from the original macros, so I guess we're relatively safe as this code was battle-tested.

However, now that we have proper functions here, I guess we can make them a bit more resilient to invalid input, like this:

```c
size_t soter_container_data_size(const soter_container_hdr_t* hdr)
{
    size_t data_size = 0;
    if (!hdr) {
        return 0;
    }
    data_size = be32toh(hdr->size);
    if (data_size < sizeof(soter_container_hdr_t)) {
        return 0;
    }
    return data_size - sizeof(soter_container_hdr_t);
}
```

What do you think?

Reply:
shouldn't we cast result to host endianness as well?

```
return be32toh(data_size - sizeof(soter_container_hdr_t));
```

Reply:
we should convert data_size to host, because it was packed as big endian. but `sizeof(...)` will return value in host endian. so `be32toh(data_size)` will be in host endian, then use `sizeof` result in host endian.
if do as you say, then we big endian `data_size` will mix with host endian `sizeof` and then convert result to big endian. it's not what we need

Reply:
okay got it

Reply:
@ilammy – i think you're right, and we should add input params checks

Reply:
I've added safety checks to the new functions (converted from macros) as well as to the old ones.